Silind-Software/direflow
# Feature/add wrapper plugins

**What does this pull request introduce? Please describe**  
Adds a new generic pattern for applying plugins.
The idea is based on a pattern where all plugins now get preregistered but applied on runtime.
Any plugin now falls into either one of three categories:
- Applies a plugin by introducing some side effect.
- Applies a plugin by wrapping the root React App. (HOC).
- Applies a plugin by wrapping the root React App thus returning `shadowChildren` that will be appended into the shadow-dom before the root React App.

This fixes #105, #106 and #94, thus greatly improves how external-scripts get applied.

Furthermore, it introduces a nice generic pattern for creating new plugins.
Any new plugin must now follow the contract of a `pluginRegistrator` type, and must be applied by the `registerPlugin` function. 

**To verify**  
1. **Create new Direflow Setup**
2. **Add styled-components, see that it works:**
   ```
   yarn add styled-components
   yarn add @types/styled-components // if TS
   ```
   ```
   plugins: [
     {
       name: 'styled-components'
     },
   ];
   ```

   ```
   import styled from 'styled-components';

   const StyledButton = styled.div`
     width: 300px;
     height: 50px;
     background-color: red;
   `;

   const App = (props) => {
     ...
     return <StyledButton>Button</StyledButton>
   }
   ```
3. **Add material-ui. See that it works:**
   ```
   yarn add @material-ui/core
   ```
   ```
   plugins: [
     {
       name: 'material-ui'
     },
   ];
   ```
   ```
   import Button from '@material-ui/core/Button';

   const App = (props) => {
     ...
     return (
       <Button variant='contained' color='primary'>
          My Material UI Button
        </Button>
     );
   }
   ```
4. **Add external script (Bootstrap), see that it works:**
   ```
   plugins: [
     {
      name: 'external-loader',
      options: {
        paths: [
          'https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/css/bootstrap.min.css'
        ],
      }
    },
   ];
   ```
   ```
   const App = (props) => {
     ...
     return <button type="button" class="btn btn-primary">Primary</button>;
   }
   ```
5. See that it all works in combination with each other.


**Version**  
3.4.5
